piker/brokers/kucoin.py

```diff
@@ -0,0 +1,668 @@
+                value = data[field_name]
+
+                match field_name:
+                    case 'index':
+                        row.append(int(value))
+                    case 'time':
+                        row.append(value)
+                    case _:
+                        row.append(float(value))
+
+            new_bars.append(tuple(row))
+
+        array = np.array(new_bars, dtype=_ohlc_dtype) if as_np else bars
+        return array
+
+
+def fqsn_to_kucoin_sym(fqsn: str, pairs: dict[str, KucoinMktPair]) -> str:
+    pair_data = pairs[fqsn]
+    return pair_data.baseCurrency + '-' + pair_data.quoteCurrency
+
+
+def kucoin_sym_to_fqsn(sym: str) -> str:
+    return sym.lower().replace('-', '')
+
+
+@acm
+async def get_client() -> AsyncGenerator[Client, None]:
+    client = Client()
+    await client.cache_pairs()
+
+    yield client
+
+
+@tractor.context
+async def open_symbol_search(
+    ctx: tractor.Context,
+) -> None:
+    async with open_cached_client('kucoin') as client:
+        # load all symbols locally for fast search
+        await client.cache_pairs()
+        await ctx.started()
+
+        async with ctx.open_stream() as stream:
+            async for pattern in stream:
+                await stream.send(await client.search_symbols(pattern))
+                log.info('Kucoin symbol search opened')
+
+
+async def stream_quotes(
+    send_chan: trio.abc.SendChannel,
+    symbols: list[str],
+    feed_is_live: trio.Event,
+    loglevel: str = None,
+    # startup sync
+    task_status: TaskStatus[tuple[dict, dict]] = trio.TASK_STATUS_IGNORED,
+) -> None:
+    '''
+    Required piker api to stream real-time data.
+    Where the rubber hits the road baby
+
+    '''
+    connect_id = str(uuid4())
+
+    async with open_cached_client('kucoin') as client:
+        log.info('Starting up quote stream')
+        # loop through symbols and sub to feedz
+        for sym in symbols:
+            token, ping_interval = await client._get_ws_token()
+            pairs = await client.cache_pairs()
+            kucoin_sym = pairs[sym].symbol
+
+            init_msgs = {
+                # pass back token, and bool, signalling if we're the writer
+                # and that history has been written
+                sym: {
+                    'symbol_info': {
+                        'asset_type': 'crypto',
+                        'price_tick_size': 0.0005,
+                        'lot_tick_size': 0.1,
+                    },
+                    'shm_write_opts': {'sum_tick_vml': False},
+                    'fqsn': sym,
+                },
+            }
+
+            @acm
+            async def subscribe(ws: wsproto.WSConnection):
+                @acm
+                async def open_ping_task(ws: wsproto.WSConnection):
+                    '''
+                    Spawn a non-blocking task that pings the ws
+                    server every ping_interval so Kucoin doesn't drop
```

Comment:
what's `ping_interval`?

Reply:
check out the docs in the docstring for `_get_ws_token`